File: php/iforevents-mixpanel/tests/MixpanelIntegrationTest.php
<?php

declare(strict_types=1);

namespace IForevents\Tests;

use IForevents\Iforevents;
use IForevents\Integrations\MixpanelIntegration;
use Mixpanel;
use PHPUnit\Framework\TestCase;

final class MixpanelIntegrationTest extends TestCase
{
    public function testForwarding(): void
    {
        $calls = [];
        $people = new class ($calls) {
            public function __construct(private array &$calls)
            {
            }

            public function set($id, $props, $ip = null, $ignoreTime = false, $ignoreAlias = false): void
            {
                $this->calls[] = ['people_set', $id, $props];
            }
        };
        $client = $this->getMockBuilder(Mixpanel::class)->disableOriginalConstructor()->onlyMethods(['track', 'flush'])->getMock();
        $client->people = $people;
        $client->method('track')->willReturnCallback(static function ($event, $props) use (&$calls): void {
            $calls[] = ['track', $event, $props];
        });
        $client->expects(self::atLeastOnce())->method('flush');

        $ife = new Iforevents([new MixpanelIntegration(client: $client)], static fn () => []);
        $ife->init();
        $ife->track('anon');
        $ife->identify('u', ['plan' => 'pro', 'nested' => ['x' => 1], 'nil' => null]);
        $ife->track('paid', ['amount' => 1]);
        $ife->page('Home', [], 'load');
        $ife->reset();
        $ife->track('again');
        $ife->shutdown();

        self::assertSame('server', $calls[0][2]['distinct_id']);
        self::assertSame(['people_set', 'u', ['plan' => 'pro', 'nested_x' => 1]], $calls[1]);
        self::assertSame('u', $calls[2][2]['distinct_id']);
        self::assertSame(1, $calls[2][2]['amount']);
        self::assertArrayHasKey('time', $calls[2][2]);
        self::assertSame('Home', $calls[3][1]);
        self::assertSame('load', $calls[3][2]['navigation_type']);
        self::assertSame('server', $calls[4][2]['distinct_id']);
    }

    public function testRealClientConstructs(): void
    {
        $integration = new MixpanelIntegration('token', options: ['consumer' => 'file', 'file' => '/tmp/iforevents-mixpanel-test.log']);
        $integration->init();
        self::assertSame('MixpanelIntegration', $integration->name);
    }
}

File: java/iforevents-mixpanel/src/test/java/com/iforevents/integrations/mixpanel/MixpanelIntegrationTest.java
package com.iforevents.integrations.mixpanel;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.iforevents.Iforevents;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.json.JSONObject;
import org.junit.jupiter.api.Test;

class MixpanelIntegrationTest {
    @Test
    void forwardsIdentifyTrackPageReset() {
        final List<JSONObject> sent = new ArrayList<JSONObject>();
        MixpanelIntegration mp = new MixpanelIntegration("token", new MixpanelIntegration.Sender() {
            @Override
            public void send(JSONObject m) {
                sent.add(m);
            }
        }, "server", null);
        Iforevents ife = Iforevents.builder().integration(mp).context(new Iforevents.ContextProvider() {
            @Override
            public Map<String, Object> get() {
                return Collections.emptyMap();
            }
        }).build();
        ife.init();
        ife.track("anon");
        Map<String, Object> traits = new LinkedHashMap<String, Object>();
        traits.put("plan", "pro");
        traits.put("nil", null);
        ife.identify("u", traits);
        ife.track("paid", Collections.singletonMap("amount", 1));
        ife.page("Home", null, "load", null, null);
        ife.reset();
        ife.track("again");
        assertEquals(5, sent.size());
        assertEquals("event", sent.get(0).getString("message_type"));
        assertEquals("server", sent.get(0).getJSONObject("message").getJSONObject("properties").getString("distinct_id"));
        JSONObject people = sent.get(1);
        assertEquals("people", people.getString("message_type"));
        assertEquals("u", people.getJSONObject("message").getString("$distinct_id"));
        assertEquals("pro", people.getJSONObject("message").getJSONObject("$set").getString("plan"));
        assertFalse(people.getJSONObject("message").getJSONObject("$set").has("nil"));
        JSONObject paid = sent.get(2).getJSONObject("message");
        assertEquals("paid", paid.getString("event"));
        assertEquals("u", paid.getJSONObject("properties").getString("distinct_id"));
        assertEquals(1, paid.getJSONObject("properties").getInt("amount"));
        assertTrue(paid.getJSONObject("properties").has("time"));
        assertEquals("Home", sent.get(3).getJSONObject("message").getString("event"));
        assertEquals("server", sent.get(4).getJSONObject("message").getJSONObject("properties").getString("distinct_id"));
    }
}
